components/org.wso2.carbon.identity.application.authenticator.biometric.servlet/src/main/java/org/wso2/carbon/identity/application/authenticator/biometric/servlet/servlet/BiometricServlet.java

```diff
@@ -0,0 +1,138 @@
+    private static final Log log = LogFactory.getLog(BiometricServlet.class);
+    private BiometricDataStoreImpl biometricDataStoreInstance = BiometricDataStoreImpl.getInstance();
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        if (!(request.getParameterMap().containsKey(BiometricServletConstants.INITIATOR))) {
+            if (log.isDebugEnabled()) {
+                log.debug("Invalid syntax as the query parameter for initiator is missing.");
+            }
+        } else { // If the initiator is not null, else block is executed.
+            String initiator = request.getParameter(BiometricServletConstants.INITIATOR);
+            if (!(BiometricServletConstants.WEB.equals(initiator) && request.getParameterMap()
+                    .containsKey(BiometricServletConstants.CONTEXT_KEY))) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Unsupported HTTP GET request or session data key is null.");
+                }
+            } else { // If the initiator is equal to WEB and if the query parameter session data
+                // key is not null, else block is executed.
+                handleWebResponse(request, response);
+            }
+        }
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        if (!request.getParameterMap().containsKey(BiometricServletConstants.INITIATOR)) {
+            PrintWriter out = response.getWriter();
+            out.println("Invalid request!");
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            response.setContentType("text/html");
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Query parameter for initiator is missing.");
+            response.getWriter().write("Invalid syntax as the query parameter for initiator is missing.");
+            response.getWriter().flush();
+            return;
+        }
+        String initiator = request.getParameter(BiometricServletConstants.INITIATOR);
+
+        if (!BiometricServletConstants.MOBILE.equals(initiator)) {
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Unsupported HTTP request from a mobile device.");
+            return;
+        }
+        handleMobileResponse(request, response);
+    }
+
+    private void handleWebResponse(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        WaitStatus waitResponse = new WaitStatus();
+        String sessionDataKeyWeb = request.getParameter(BiometricServletConstants.CONTEXT_KEY);
+        String signedChallengeExtracted = biometricDataStoreInstance.getSignedChallenge(sessionDataKeyWeb);
+        if (StringUtils.isEmpty(signedChallengeExtracted)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Signed challenge sent from the mobile application is null.");
+            }
+
```

Comment:
Why do we have only a debug log here? Aren't we handling the case where signedChallengeExtracted is null?

Reply:
It wont be a null at any time according to the logic in the servlet. If both the CONTEXT_KEY and CHALLENGE from the mobile are not null only , the biometricDataStore hashmap will be updated. If even one of the parameters are missing, its handled in the 1st if condition of handleMobileResponse() method.